RavenDB-19254 - Sharding - duplicated conflicts

shiranshalom opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19254/Sharding-duplicated-conflicts

### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

shiranshalom committed:
RavenDB-19254 - Sharding - duplicated conflicts

ppekrol commented:
It does not compile?

karmeli87 commented on test/SlowTests/Sharding/Client/Operations/GetDocumentConflictsTests.cs:
```diff
@@ -60,13 +60,13 @@ public async Task CanGetReplicationDocumentConflicts()
 
-                    documentsConflict = await store.Maintenance.SendAsync(new GetConflictsOperation(start: 12));
-                    Assert.Equal(8, documentsConflict.Results.Length);
+                    documentsConflict = await store.Maintenance.SendAsync(new GetConflictsOperation(start: 2));
```
`pageSize` change I understand, but why the `start` reduced to `2` from `12`?

shiranshalom replied:
because now we have 10 unique results instead of 20

karmeli87 commented on src/Raven.Server/Documents/ConflictsStorage.cs:
```diff
@@ -110,30 +110,26 @@ internal GetConflictsPreviewResult GetConflictsPreviewResult(DocumentsOperationC
-            var conflicts = new List<GetConflictsPreviewResult.ConflictPreview>();
-            foreach (var tvr in table.SeekBackwardFromLast(ConflictsSchema.FixedSizeIndexes[AllConflictedDocsEtagsSlice], skip))
+            var conflictsDictionary = new Dictionary<string, GetConflictsPreviewResult.ConflictPreview>();
+            foreach (var tvr in table.SeekBackwardFromLast(ConflictsSchema.FixedSizeIndexes[AllConflictedDocsEtagsSlice]))
```
you will scan here through the entire table.
1. we need page size
2. we need skip

Skip is not the number of the unique Ids, but should be the number of scanned items.

shiranshalom replied:
Done

shiranshalom committed:
RavenDB-19254 - fixing PR comments#1

shiranshalom commented:
@ml054 Changes required in the Studio:
Calculating next `Start`:
`Start` = **Sum**(`Results.ScannedResults`) + 1

karmeli87 commented on src/Raven.Server/Documents/Sharding/Handlers/Processors/Replication/ShardedReplicationHandlerProcessorForGetConflicts.cs:
```diff
@@ -59,18 +58,27 @@ public ShardedGetReplicationConflictsOperation(ShardedDatabaseRequestHandler han
+                        break;
+
+                    var shard = res.Shard;
+                    _token.Pages[shard].Start += (int)res.Item.ScannedResults;
```
is `+=` correct? if you collect 2 results here next time you will skip here more than intended, no?.

shiranshalom replied:
You are right. Done.

karmeli87 commented on src/Raven.Server/Documents/Handlers/Processors/Replication/AbstractReplicationHandlerProcessorForGetConflicts.cs:
```diff
@@ -32,29 +31,27 @@ public override async ValueTask ExecuteAsync()
                         [nameof(GetConflictsPreviewResult.ConflictPreview.Id)] = conflict.Id,
                         [nameof(GetConflictsPreviewResult.ConflictPreview.LastModified)] = conflict.LastModified,
-                        [nameof(GetConflictsPreviewResult.ConflictPreview.ConflictsPerDocument)] = conflict.ConflictsPerDocument
+                        [nameof(GetConflictsPreviewResult.ConflictPreview.ConflictsPerDocument)] = conflict.ConflictsPerDocument,
```
can we have `ToJson` on the `conflict` ?

shiranshalom replied:
done

ml054 commented:
@shiranshalom I can see we added continuation token some time ago. So why I need to provide some special params as well? Or we removed continuation token?

shiranshalom commented:
> @shiranshalom I can see we added continuation token some time ago. So why I need to provide some special params as well? Or we removed continuation token?

@ml054 We still have `ContinuationToken` :-). 
We need `Results.ScannedResults` for regular case (to calculate `Start`).

shiranshalom committed:
RavenDB-19254 - fixing PR comments

ml054 commented:
@shiranshalom let's move scanned results to top level (next to total results field) to mimic behavior we have in query results.

ml054 committed:
RavenDB-19254 - Sharding - duplicated conflicts - studio part

ml054 commented:
@shiranshalom I did my changes in blind - so either please test studio paging for sharded and non-sharded dbs or let me know once you finish your changes and I will test.

shiranshalom committed:
RavenDB-19254 - fixing PR comments#3

shiranshalom commented:
> @shiranshalom let's move scanned results to top level (next to total results field) to mimic behavior we have in query results.

@ml054 Done

shiranshalom commented:
> @shiranshalom I did my changes in blind - so either please test studio paging for sharded and non-sharded dbs or let me know once you finish your changes and I will test.

Tested manually for both Sharded and Non-sharded.
